feat(state-middleware): implementation

paul-spence-amido opened the pull request.

spenc259 committed:
feat(state-middleware): implementation

velvetkevorkian commented on libs/framework/middlewares/src/initialState.tsx:
```diff
@@ -0,0 +1,13 @@
+export const stateMiddleware = (state) => {
+    const middlewareState = state
+  
+    // do we need to return an async func here?
```
No, it doesn't have to be async

velvetkevorkian commented on libs/framework/middlewares/src/prerender.tsx:
```diff
@@ -85,13 +85,17 @@ export const prerenderMiddleware = remoteEntry => {
           res.contentType('text/plain')
           res.write(stringifiedChunks)
           res.write(constants.SERIALISED_RESPONSE_SEPARATOR)
+          res.write(initialState)
```
I think we probably want this to be after `pipe(res)` so it's not a breaking change for consumers who are splitting the response to an array and destructuring it, e.g. 
```
      const [chunks, html] = res.data.split(constants.SERIALISED_RESPONSE_SEPARATOR)
```

spenc259 committed:
refactor(pr amends): switch the order of where we place the initial state in the stream and tidy up

spenc259 committed:
feat(initial state): pass the initial state from MFE to AppShell

paul-spence-amido commented on apps/AppShell/src/utils/combineInitialStates/index.ts:
```diff
@@ -0,0 +1,21 @@
+  const keys = Object.keys(initalState)
+  allstates[keys[0]] = initalState[keys[0]]
+
+  const combinedStates = Array.from(state).reduce((p: any, c: any) => {
```
100% and especially if we are going to filter the NO STATE's in federate component

Commit:
Update libs/framework/federate-component/src/server.tsx

Co-authored-by: Kyle Macquarrie <kyle.macquarrie@amido.com>

spenc259 committed:
refactor(combineinitialstates): use for each and simplyfy the logic

we now filter the state in federate component so we can remove the unecessary check for no state and
use a more readable for each
